src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelMetrics.cs

```diff
@@ -0,0 +1,240 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Connections.Features;
+using Microsoft.Extensions.Metrics;
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure;
+
+internal sealed class KestrelMetrics
+{
+    // Note: Dot separated instead of dash.
+    public const string MeterName = "Microsoft.AspNetCore.Server.Kestrel";
+
+    private readonly Meter _meter;
+    private readonly UpDownCounter<long> _currentConnectionsCounter;
+    private readonly Histogram<double> _connectionDuration;
+    private readonly Counter<long> _rejectedConnectionsCounter;
+    private readonly UpDownCounter<long> _queuedConnectionsCounter;
+    private readonly UpDownCounter<long> _queuedRequestsCounter;
+    private readonly UpDownCounter<long> _currentUpgradedRequestsCounter;
+    private readonly Histogram<double> _tlsHandshakeDuration;
+    private readonly UpDownCounter<long> _currentTlsHandshakesCounter;
+
+    public KestrelMetrics(IMeterFactory meterFactory)
+    {
+        _meter = meterFactory.CreateMeter(MeterName);
+
+        _currentConnectionsCounter = _meter.CreateUpDownCounter<long>(
+           "current-connections",
+            description: "Current connections");
+
+        _connectionDuration = _meter.CreateHistogram<double>(
+            "connection-duration",
+            description: "Connection duration");
+
+        _rejectedConnectionsCounter = _meter.CreateCounter<long>(
+           "rejected-connections",
+            description: "Rejected connections");
+
+        _queuedConnectionsCounter = _meter.CreateUpDownCounter<long>(
+           "queued-connections",
+            description: "Queued connections");
```

Comment:
Is this "things that have ever been queued" or "things that are currently in queue". If it is the former, use Counter instead of UpDownCounter. Also consider adding clarification in the description.

Reply:
It's things that are currently in queue. I'll improve the descriptions across all the counters.

Reply:
Improved descriptions.